### PR TITLE
[core] Deflakey ray syncer test due to race condition.

### LINKED ISSUE #41133: [CI] `linux://src/ray/common/test:ray_syncer_test` is failing/flaky on master.
- 531cea8880c463b1a6c91166698c74ca1603d70b FAILED [Buildkite :ray: core: cpp tests](https://buildkite.com/ray-project/postmerge/builds/1652#018bcaec-b2aa-42cc-a8d1-f032c9d08aa2)

....
Generated from flaky test tracker. Please do not edit the signature in this section.
DataCaseName-linux://src/ray/common/test:ray_syncer_test-END
....

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
The test is flakey because of the race condition. In the code we break the transaction into several parts for performance but it causes some issues.

1. Disconnect is requested by the user => RPC terminates async
2. Read/Write arrived => Read next is requested async
3. 1 and 2 has race condition. The ordering matters. If Read next happens after RPC terminates, it'll crash.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/ray_syncer/ray_syncer-inl.h [hunk 1]
```diff
@@ -205,7 +206,7 @@ class RaySyncerBidiReactorBase : public RaySyncerBidiReactor, public T {
     return false;
   }
 
-  virtual ~RaySyncerBidiReactorBase() {}
+  virtual ~RaySyncerBidiReactorBase() = default;
```
Comment (rickyyx) [inline src/ray/common/ray_syncer/ray_syncer-inl.h:209]: is this to fix the pure virtual method called issue?

**Review by rickyyx — APPROVED** [review]
So things are still single-threaded thus the shared_ptr to bool, rather than something like atomic?

**Comment by iycheng** [comment]
@rickyyx  it's still single thread. There are two threads:
- gRPC thread
- Ray thread

All work modifying the context is in Ray thread, like checking whether it's disconnected.

Managing RPC lifecycle is done in gRPC thread. Switching between the thread is the place where things go wrong.

I definitely feel there is a huge gap to make these two work efficiently and and safely. This means the API in gRPC is not easy to use and we need some wrapper on it.

**Comment by iycheng** [comment]
test draining failed. checking

https://buildkite.com/ray-project/premerge/builds/13314#018c225a-025f-49e5-9ae9-fefb630417ba

**Pull request merged**